src/v/cluster/partition.cc

```diff
@@ -152,6 +152,104 @@ ss::future<std::vector<rm_stm::tx_range>> partition::aborted_transactions_cloud(
     return _cloud_storage_partition->aborted_transactions(offsets);
 }
 
+cluster::cloud_storage_mode partition::get_cloud_storage_mode() const {
+    const auto& cfg = _raft->log_config();
+
+    if (cfg.is_read_replica_mode_enabled()) {
+        return cluster::cloud_storage_mode::read_replica;
+    }
+    if (cfg.is_tiered_storage()) {
+        return cluster::cloud_storage_mode::full;
+    }
+    if (cfg.is_tiered_storage()) {
+        return cluster::cloud_storage_mode::full;
+    }
+    if (cfg.is_archival_enabled()) {
+        return cluster::cloud_storage_mode::write_only;
+    }
+    if (cfg.is_remote_fetch_enabled()) {
+        return cluster::cloud_storage_mode::read_only;
+    }
+
+    return cluster::cloud_storage_mode::disabled;
+}
+
+partition_cloud_storage_status partition::get_cloud_storage_status() const {
+    auto wrap_model_offset =
+      [this](model::offset o) -> std::optional<kafka::offset> {
+        if (o == model::offset{}) {
+            return std::nullopt;
+        }
+        return model::offset_cast(
+          get_offset_translator_state()->from_log_offset(o));
+    };
+
+    auto wrap_ts =
+      [](ss::lowres_clock::time_point tp) -> std::optional<model::timestamp> {
+        if (tp.time_since_epoch().count() == 0) {
+            return std::nullopt;
+        }
+
+        return model::timestamp{
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+            tp.time_since_epoch())
```

Comment:
looks good. the only consideration left with lowres_clock would be that it's only going to be able to give meaningful durations when both time points involved in the subtraction have been sampled since the redpanda process started and ideally on the same core. i took a quick peek and it didn't look like any of the non-`now()` values were restored from disk or received from rpc, but it wasn't a thorough investigation.

Reply:
The original time points are in memory-only (they don't get restored), so that should be fine.

Reply:
cool, thanks